Ruler Tool -- initial placement [GEOCODE-35]

kswenson opened the pull request:
[[GEOCODE-35](https://concord-consortium.atlassian.net/browse/GEOCODE-35)] Ruler Tool -- initial placement

A lot of subtleties here. A few highlights:
- When entering ruler mode the map animates to a top view with no vertical exaggeration.
- When exiting ruler mode the map animates back to its original pitch and vertical exaggeration.
- After the first point is placed the cursor is hidden and a cesium entity on the map simulates the cursor.
- The positions of the points and of the connecting line are computed dynamically using cesium's `CallbackProperty` and `CallbackPositionProperty` classes, so that they don't need to be destroyed/recreated on position changes.
- The label is rendered as a standard React component that is placed geometrically near the midpoint of the connecting line. The placement code is subtle.
- The story doesn't specify whether the distance is meant to be the 3D distance (i.e. taking elevation into account) or the 2D distance. For now I've chosen the 3D distance since the major purpose seems to be to answer the question how far would the lava have to travel to get to a given point, in which case the elevation is relevant. (It's also slightly easier to calculate, but it would also be straightforward to compute it the other way.)

[GEOCODE-35]: https://concord-consortium.atlassian.net/browse/GEOCODE-35?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

tealefristoe commented:
I just started looking at the next PR and noticed you already fixed the issue where the label does not stay with the line when the camera is moved. Nice!

kswenson commented:
> Overall looks good 👍 I have a good handful of suggestions and questions, but you can decide which of them are worth addressing before merging.
> 
> * The biggest issue in my mind is that the label does not stay with line when moving the camera. It also feels a bit weird that you can pan/rotate the camera while dropping the ruler, but it's probably the correct behavior.

Already fixed in #429.

> * When you press the ruler button repeatedly in quick succession, you lose initial pitch. It's kind of dumb to repeatedly push the ruler button, so maybe the user deserves this, but I bet kids will do it.

Fixed by additional commit in #429 by disabling button during animation.

> * I suspect the distance should be the 2d distance. It makes sense given that we're in top-down mode. Plus the 3d distance is pretty misleading if there are peaks or valleys between the points.

After discussion in the demo/planning meeting, we're leaving it the 3D distance. The 3D distance doesn't take intervening terrain into account, only the straight-line distance between the two points (including elevation). Given two points at sea level on opposite coasts, the 2D and 3D distances would essentially be the same, regardless of whether there's a peak in between.

> * Of the inline comments I left, the one about changing `currVerticalExaggeration` to `verticalExaggeration` is the one I'd most recommend. I think the vast majority of clients (all?) will expect the current vertical exaggeration, and having `verticalExaggeration` sometimes not return this will create an opportunity for clients to make confusing mistakes.

I agree, and I was planning to do something like that but then I got a little worried about backward compatibility with existing model configurations, but maybe we just won't worry about that right now. Fixed by additional commit in #429.

kswenson commented on src/stores/ui-store.ts:
```diff
@@ -83,7 +91,14 @@ const UIStore = types.model("UI", {
+  rulerLine: undefined as ILavaCoderRulerLine | undefined
+}))
 .views((self) => ({
+  get currVerticalExaggeration() {
```
I agree, and I was planning to do something like that but then I got a little worried about backward compatibility with existing saved model configurations, but maybe we just won't worry about that right now. Fixed by additional commit in #429.

kswenson commented on src/components/lava-coder/use-camera-controls.ts:
```diff
@@ -39,6 +42,55 @@ export function useCameraControls(viewer: CesiumWidget | null, verticalExaggerat
+        window.clearInterval(interval);
+        animationInterval.current = null;
+        currentPitch = pitch;
+        uiStore.setTempVerticalExaggeration(exaggeration === 1 ? exaggeration : undefined);
```
Fixed in additional commit in #429.

kswenson commented on src/components/lava-coder/use-ruler-mode.ts:
```diff
@@ -0,0 +1,167 @@
+  useEffect(() => {
+    if (!viewer) return;
+
+    if (getFirstPoint() && !firstPointEntity.current) {
+      firstPointEntity.current = viewer.entities.add({
+        id: kFirstPointId,
+        position: new CallbackPositionProperty(() => getFirstPoint()!, false),
```
I added a comment.

kswenson commented on src/components/lava-coder/use-ruler-mode.ts:
```diff
@@ -0,0 +1,167 @@
+      connectingLineEntity.current = viewer.entities.add({
+        id: kConnectingLineId,
+        polyline: {
+          positions: new CallbackProperty(() => [getFirstPoint()!, getSecondPoint()!], false),
```
I added a comment.

kswenson commented on src/components/lava-coder/use-ruler-mode.ts:
```diff
@@ -0,0 +1,167 @@
+        }
+      });
+    }
+    else if (!getSecondPoint() && secondPointEntity.current) {
```
In the code for setting the points, what matters is whether the points have been placed yet, so `...Static.current` is the most relevant thing to test. For code that is related to the connecting line, `getSecondPoint()` returns either the statically defined point or the dynamically defined one, so it's the relevant metric. For `getFirstPoint()`, it doesn't matter which is used, but I defined so the code for the two points parallel each other better.
